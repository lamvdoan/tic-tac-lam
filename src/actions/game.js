import { UPDATE_GRID_ELEMENT, SET_IS_GAME_ACTIVE, SET_CURRENT_PLAYER } from '../properties/actions';

export function cellIsMarked(index) {
    return {
        type: UPDATE_GRID_ELEMENT,
        index
    };
}

export function setIsGameActive() {
    return {
        type: SET_IS_GAME_ACTIVE,
        isGameActive: false
    };
}

export function setCurrentPlayer() {
    return {
        type: SET_CURRENT_PLAYER
    };
}
