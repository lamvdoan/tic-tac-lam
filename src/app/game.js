import { Player } from './player';
import { UNOCCUPIED_CELL, MAX_NUMBER_OF_CELLS, PLAYER_ONE_ID, PLAYER_TWO_ID, WINNING_CELLS_LIST } from './properties';

export class Game {
    constructor() {
        this.players = [new Player('X', PLAYER_ONE_ID), new Player('O', PLAYER_TWO_ID)];
    }

    startGame() {
        this.grid = Array(MAX_NUMBER_OF_CELLS).fill(UNOCCUPIED_CELL);
        this.isGameActive = true;
        this.currentPlayer = this.players[0];
        console.log('Game starting...');
    }

    chooseACell(index) {
        if (this.isGameActive) {
            if (this.isMoveValid(index)) {
                this.cellIsMarked(index);
                this.checkForWinCondition();
                this.checkForCatCondition();
                this.nextPlayer();
            } else {
                // Display message to try again
            }
        }
    }

    isMoveValid(index) {
        return (this.grid[index] === UNOCCUPIED_CELL) ? true : false;
    }

    cellIsMarked(index) {
        this.grid[index] = this.currentPlayer.getMarker();
    }

    checkForWinCondition() {
        const getPlayerMarkedCellsReducer = (accumulator, cell, index) => {
            if (cell === this.currentPlayer.getMarker()) {
                accumulator.push(index);
            }

            return accumulator;
        };

        const playerCells = this.grid.reduce(getPlayerMarkedCellsReducer, []);

        for (let i = 0; i < WINNING_CELLS_LIST.length; i++) {
            let winningCells = Array.from(WINNING_CELLS_LIST[i]);
            let didPlayerWin = winningCells.every(cell => playerCells.includes(cell));

            if (didPlayerWin) {
                this.isGameActive = false;
                break;

            // display victory for specified player
            }
        }
    }

    checkForCatCondition() {
        if (this.isGameActive) {
            const countOfOccupiedCells = this.grid.filter(cell => cell != UNOCCUPIED_CELL);

            if (countOfOccupiedCells.length == MAX_NUMBER_OF_CELLS) {
                this.isGameActive = false;

                // display cat condition
            }
        }
    }

    nextPlayer() {
        if (this.isGameActive) {
            this.currentPlayer = (this.currentPlayer.getPlayerId() === PLAYER_ONE_ID) ? this.players[1] : this.players[0];

            // display whose turn
        }
    }
}
