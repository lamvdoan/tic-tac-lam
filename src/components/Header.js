import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => (
    <header className="header">
        <div className="content-container">
            <Link className="header__title" to="/">
                <h1>Tic Tac Lam</h1>
            </Link>
        </div>
    </header>
);
export default Header;
