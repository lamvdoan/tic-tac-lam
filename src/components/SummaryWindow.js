import React from 'react';
import { connect } from 'react-redux';

export const SummaryWindow = ({ currentPlayer }) => (
    <div>
        <p>Current Player: {currentPlayer}</p>
    </div>
);

const mapStateToProps = (state) => ({
    currentPlayer: state.currentPlayer
});

export default connect(mapStateToProps)(SummaryWindow);