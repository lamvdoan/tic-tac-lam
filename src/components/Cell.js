import React from 'react';
import { connect } from 'react-redux';
import { cellIsMarked, setIsGameActive, setCurrentPlayer } from '../actions/game';
import { O_CELL_IMAGE, X_CELL_IMAGE, EMPTY_CELL_IMAGE, IMAGE_HOME } from '../properties/imageFileSystem';

const imagePath = {
    emptyCell: `${IMAGE_HOME}/${EMPTY_CELL_IMAGE}`,
    oCell: `${IMAGE_HOME}/${O_CELL_IMAGE}`,
    xCell: `${IMAGE_HOME}/${X_CELL_IMAGE}`,
}

export class Cell extends React.Component {
    onClick = () => {
        console.log('Clicking cell: ' + this.props.id);
        this.props.cellIsMarked(this.props.id);
        this.props.setCurrentPlayer();

        // console.log('Grid Letter: ' + this.props.game.grid[this.props.id]);
        // console.log('Grid: ' + this.props.game.grid);
    };

    getImageName = () => {
        if (this.props.game.grid[this.props.id] === 'X') {
            return 'xCell';
        } else if (this.props.game.grid[this.props.id] === 'O') {
            return 'oCell';
        }
    }

    render() {
        const imageName = this.getImageName();
        return (
            <img onClick={this.onClick} className="cell" src={imagePath[imageName]} />
        );
    }
}

const mapStateToProps = (state) => ({
    game: state.game
});

const mapDispatchToProps = (dispatch) => ({
    cellIsMarked: (index) => dispatch(cellIsMarked(index)),
    setIsGameActive: () => dispatch(setIsGameActive()),
    setCurrentPlayer: () => dispatch(setCurrentPlayer())
});

export default connect(mapStateToProps, mapDispatchToProps)(Cell);
