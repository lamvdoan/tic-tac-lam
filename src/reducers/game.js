import { Player } from '../app/player';
import { 
    MAX_NUMBER_OF_CELLS, 
    UNOCCUPIED_CELL, 
    PLAYER_ONE_ID, 
    PLAYER_TWO_ID, 
    X_CELL, 
    O_CELL
} from '../properties/game';

import { 
    RESTART_GAME, 
    UPDATE_CELL_ON_BOARD, 
    SET_PLAYER_WIN_STATUS, 
    GET_NEXT_PLAYER,
    SET_CAT_STATUS
} from '../actions/types';

const players = [
    new Player(X_CELL, PLAYER_ONE_ID), 
    new Player(O_CELL, PLAYER_TWO_ID)
];

const gameReducerDefaultState = {
    grid: Array(MAX_NUMBER_OF_CELLS).fill(UNOCCUPIED_CELL),
    isGameActive: true,
    didPlayerWinGame: false,
    players: players,
    currentPlayer: players[0]
};

export default (state = gameReducerDefaultState, action) => {
    switch (action.type) {
        case RESTART_GAME:
            return {
                ...gameReducerDefaultState
            };
        case UPDATE_CELL_ON_BOARD:
            return {
                ...state, 
                grid: state.grid.map((current, index) => {
                    if (index == action.index) {
                        return state.currentPlayer.getMarker();
                    }

                    return current;
                })
            };
        case SET_PLAYER_WIN_STATUS:
            return {
                ...state, 
                isGameActive: false,
                didPlayerWinGame: true
            };
        case SET_CAT_STATUS:
            return {
                ...state,
                isGameActive: false
            }
        case GET_NEXT_PLAYER:
                return {
                    ...state, 
                    currentPlayer: state.currentPlayer.getPlayerId() === state.players[0].getPlayerId() ? state.players[1] : state.players[0]
                };
        default:
            return state;
    }
};
