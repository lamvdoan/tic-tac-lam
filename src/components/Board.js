import React from 'react';
import Cell from './Cell';

export class Board extends React.Component {
    render() {
        return (
            <div className="board">
                <div className="board__row">
                    <Cell id="0" />
                    <Cell id="1"/>
                    <Cell id="2"/>
                </div>
                <div className="board__row">
                    <Cell id="3"/>
                    <Cell id="4"/>
                    <Cell id="5"/>
                </div>
                <div className="board__row">
                    <Cell id="6"/>
                    <Cell id="7"/>
                    <Cell id="8"/>
                </div>
            </div>
        );
    }
}

export default Board;
