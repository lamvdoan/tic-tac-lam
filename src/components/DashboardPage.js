import React from 'react';
import Board from './Board';

const DashboardPage = () => (
    <div>
        <Board />
    </div>
);

export default DashboardPage;