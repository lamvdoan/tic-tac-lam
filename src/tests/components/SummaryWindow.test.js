import React from 'react';
import { shallow } from 'enzyme';
import { SummaryWindow } from '../../components/SummaryWindow';
import { gameReducerDefaultState, alternateGameReducerState } from '../fixtures/game';

test('should render summary window with player 1\'s turn', () => {
    const game = {
        ...gameReducerDefaultState,
    }
    const wrapper = shallow(<SummaryWindow game={game} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render summary window with player 2\'s turn', () => {
    const game = {
        ...alternateGameReducerState,
    }
    const wrapper = shallow(<SummaryWindow game={game} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render summary window with player 1 winning', () => {
    const game = {
        ...gameReducerDefaultState,
        isGameActive: false,
        didPlayerWinGame: true
    }
    const wrapper = shallow(<SummaryWindow game={game} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render summary window with player 2 winning', () => {
    const game = {
        ...alternateGameReducerState,
        isGameActive: false,
        didPlayerWinGame: true
    }
    const wrapper = shallow(<SummaryWindow game={game} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render summary window with cat', () => {
    const game = {
        ...alternateGameReducerState,
        isGameActive: false,
        didPlayerWinGame: false
    }
    const wrapper = shallow(<SummaryWindow game={game} />);
    expect(wrapper).toMatchSnapshot();
});
