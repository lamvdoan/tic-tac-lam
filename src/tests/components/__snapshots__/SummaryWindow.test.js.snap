// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render summary window with cat 1`] = `
<div
  className="summary-window"
>
  <p>
    Cat!
  </p>
</div>
`;

exports[`should render summary window with player 1 winning 1`] = `
<div
  className="summary-window"
>
  <p>
    Player 1 Wins!
  </p>
</div>
`;

exports[`should render summary window with player 1's turn 1`] = `
<div
  className="summary-window"
>
  <p>
    Player 1's Turn
  </p>
</div>
`;

exports[`should render summary window with player 2 winning 1`] = `
<div
  className="summary-window"
>
  <p>
    Player 2 Wins!
  </p>
</div>
`;

exports[`should render summary window with player 2's turn 1`] = `
<div
  className="summary-window"
>
  <p>
    Player 2's Turn
  </p>
</div>
`;
